File: typescript/src/metadata.ts
// Generator metadata helpers. Mirrors specs/acceptance-generator.md.
import { createHash } from "node:crypto";
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";

export const SCHEMA_VERSION = 1;

export function metadataFilename(featurePath: string): string {
  const lowered = featurePath.toLowerCase();
  const hyphenated = lowered.replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
  return `${hyphenated}.json`;
}

export function implementationHash(generatedFiles: string[]): string {
  const sorted = [...generatedFiles].sort();
  const h = createHash("sha256");
  for (const file of sorted) {
    h.update(file);
    h.update(Buffer.from([0]));
    h.update(readFileSync(file));
    h.update(Buffer.from([0]));
  }
  return `sha256:${h.digest("hex")}`;
}

export interface MetadataInput {
  metadataPath: string;
  featurePath: string;
  irPath: string;
  generatedFiles: string[];
}

export function writeMetadata(input: MetadataInput): void {
  mkdirSync(dirname(input.metadataPath), { recursive: true });
  const payload = {
    schema_version: SCHEMA_VERSION,
    feature_path: input.featurePath,
    ir_path: input.irPath,
    implementation_hash: implementationHash(input.generatedFiles),
    hash_scope: "generated_files",
    generated_files: input.generatedFiles,
  };
  writeFileSync(input.metadataPath, JSON.stringify(payload, null, 2) + "\n", "utf-8");
}

File: rust/aps-kit/src/metadata.rs
use std::fs;
use std::path::{Path, PathBuf};

use sha2::{Digest, Sha256};

pub const SCHEMA_VERSION: u32 = 1;

pub fn metadata_filename(feature_path: &str) -> String {
    let lower = feature_path.to_lowercase();
    let mut out = String::with_capacity(lower.len());
    let mut last_dash = true; // strip leading dashes
    for ch in lower.chars() {
        if ch.is_ascii_alphanumeric() {
            out.push(ch);
            last_dash = false;
        } else if !last_dash {
            out.push('-');
            last_dash = true;
        }
    }
    while out.ends_with('-') {
        out.pop();
    }
    out.push_str(".json");
    out
}

pub fn implementation_hash(generated_files: &[PathBuf]) -> Result<String, String> {
    let mut sorted: Vec<&PathBuf> = generated_files.iter().collect();
    sorted.sort();
    let mut hasher = Sha256::new();
    for path in sorted {
        let path_str = path.to_string_lossy();
        hasher.update(path_str.as_bytes());
        hasher.update([0u8]);
        let bytes = fs::read(path).map_err(|e| format!("hash {path:?}: {e}"))?;
        hasher.update(&bytes);
        hasher.update([0u8]);
    }
    Ok(format!("sha256:{:x}", hasher.finalize()))
}

pub struct MetadataInput<'a> {
    pub metadata_dir: &'a Path,
    pub feature_path: &'a str,
    pub ir_path: &'a str,
    pub generated_files: &'a [PathBuf],
}

pub fn write_metadata(input: MetadataInput<'_>) -> Result<PathBuf, String> {
    fs::create_dir_all(input.metadata_dir).map_err(|e| e.to_string())?;
    let hash = implementation_hash(input.generated_files)?;
    let payload = serde_json::json!({
        "schema_version": SCHEMA_VERSION,
        "feature_path": input.feature_path,
        "ir_path": input.ir_path,
        "implementation_hash": hash,
        "hash_scope": "generated_files",
        "generated_files": input.generated_files.iter().map(|p| p.to_string_lossy().to_string()).collect::<Vec<_>>(),
    });
    let out = input.metadata_dir.join(metadata_filename(input.feature_path));
    let serialized = serde_json::to_string_pretty(&payload).map_err(|e| e.to_string())?;
    fs::write(&out, format!("{serialized}\n")).map_err(|e| e.to_string())?;
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalizes_filenames() {
        assert_eq!(
            metadata_filename("features/Hunt The Wumpus.feature"),
            "features-hunt-the-wumpus-feature.json"
        );
        assert_eq!(
            metadata_filename("features/orders/Cancel Order.feature"),
            "features-orders-cancel-order-feature.json"
        );
        assert_eq!(
            metadata_filename("Features/API v2/Happy Path.feature"),
            "features-api-v2-happy-path-feature.json"
        );
    }
}
